### PR TITLE
Implemented backoff for 429 http status

### LINKED ISSUE #8: Sometimes milv send too many request and got 429 as resposne
**Description**
Milv sometimes call external service too intensively

**Expected result**
Milv should retried the call when got 429 status code after some time.

**Actual result**
Milv is flaky in CI pipelines, because it sends too many request.

### PR DESCRIPTION
**Pull request by dbadura**
**Description**

Changes proposed in this pull request:

- add backoff functionality for 429 status code

**Related issue(s)**
#8

### CODE REVIEW

File: pkg/config_file.go [hunk 1]
```diff
@@ -129,6 +133,14 @@ func getDefaultIntIfNil(defaultValue int, value *int) int {
 	return *value
 }
 
+func getDefaultDurationIfProvided(globalCfg, fileCfg time.Duration) time.Duration {
+	var backoff = globalCfg
+	if fileCfg != 0 {
+		backoff = fileCfg
+	}
+	return backoff
+}
```
Comment (tgorgol) [inline pkg/config_file.go]: In other functions it is:
> get default if nil

here it is:
> get default if provided

and yet the logic is the same, I think we should stay consistent and use "not provided" instead.

```suggestion
func getDefaultDurationIfNotProvided(defaultValue, value time.Duration) time.Duration {
	if value != 0 {
		return value
	}
	return defaultValue
}
```

File: pkg/file.go [hunk 2]
```diff
@@ -37,6 +39,13 @@ func NewFile(filePath string, fileLinks Links, config *FileConfig) (*File, error
 		return nil, err
 	}
 
+	client := http.Client{}
+	backoff := 1 * time.Second
+	if config != nil {
+		backoff = config.Backoff
+	}
```
Comment (tgorgol) [inline pkg/file.go]: This is the second place where we're setting the default value for backoff. It's confusing and if I'm not mistaken this default will never actually be used. I assume you did this because you had to check for nil but this config will never actually be nil. Maybe we could just refactor the configs passed in the program to not be unnecessary pointers?

File: pkg/retry_test.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	before := time.Now()
+	expected := before.Add(backoff)
+	//WHEN
+	retry.Limit()
+
+	//THEN
+	afterExecution := time.Now()
+	assert.True(t, afterExecution.After(expected))
+	assert.WithinDuration(t, expected, afterExecution, 250*time.Millisecond)
+}
```
Comment (tgorgol) [inline pkg/retry_test.go]: We are basically testing the `time.Sleep` function here, I don't think it's necessary. Also `time.Sleep` doesn't guarantee a precise delay duration (there are even some OS specific edge cases) so it might result in a flaky test.

I would change the assertion to verify if the elapsed time was equal to or greater than the expected delay.

File: pkg/retry.go [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package pkg
+
+import "time"
+
+type limiter struct {
+	d time.Duration
+}
+
+func NewLimiter(d time.Duration) *limiter {
```
Comment (tgorgol) [inline pkg/retry.go]: Why `Limiter`? I think `Waiter` or `Sleeper` would be easier to understand but see my other comment about the `Validator` struct first, please.

File: pkg/validation.go [hunk 5]
```diff
@@ -12,10 +12,20 @@ import (
...
 
-type Validation struct {
+type Limiter interface {
+	Limit()
 }
 
-func (v *Validation) Links(links []Link, optionalHeaders ...Headers) []Link {
+type Validator struct {
+	client  http.Client
+	limiter Limiter
```
Comment (tgorgol) [inline pkg/validation.go]: I wonder if we even need a separate structure for this Limiter? Wouldn't a field storing the duration be sufficient? We could just call `time.Sleep()` directly and avoid unit testing `time.Sleep` function in Limiter tests.

File: pkg/validation.go [hunk 6]
```diff
@@ -147,6 +156,11 @@ func (*Validation) externalLink(link Link) (Link, error) {
 
 			resp.Body.Close()
 			break
+		} else if resp.StatusCode == http.StatusTooManyRequests {
+			status = false
+			message = "Too many requests"
+			v.limiter.Limit()
+			continue
```
Comment (tgorgol) [inline pkg/validation.go:164]: You forgot about closing the response body. Please refactor the code to use `defer` right after opening the body instead.

File: pkg/validation.go [hunk 7]
```diff
@@ -147,6 +156,11 @@ func (*Validation) externalLink(link Link) (Link, error) {
 
 			resp.Body.Close()
 			break
+		} else if resp.StatusCode == http.StatusTooManyRequests {
+			status = false
+			message = "Too many requests"
+			v.limiter.Limit()
```
Comment (tgorgol) [inline pkg/validation.go]: Maybe we could try using the `Retry-After` header if provided, and fallback to default backoff otherwise?

**Review by tgorgol — COMMENTED** [review]
Good job, few suggestions from my side.

Also, could you please open a PR to test-infra with image bump and include [this pjtester config](https://github.com/kyma-project/test-infra/blob/d9cba9c363b76544e8a277f0c80cb12ccff628b5/vpath/pjtester.yaml).

File: pkg/validation.go (hunk 5 above)
Comment (dbadura) [reply pkg/validation.go]: I think storing the duration is not sufficient.
The problem is that `time.Sleep` is static method. In unit test I want to know if the sleep was called without calling sleep itself.

File: pkg/file.go (hunk 2 above)
Comment (dbadura) [reply pkg/file.go]: I am afraid that such refactor will consume too many human lives.

I will try!

File: pkg/retry_test.go (hunk 3 above)
Comment (dbadura) [reply pkg/retry_test.go]: I agree with you that's why I use delta of 0,25 seconds in seconds assertion to be sure that I used correct duration.

You propose to left this assertion?:
```
assert.True(t, afterExecution.After(expected)) 
```

File: pkg/validation.go (hunk 7 above)
Comment (dbadura) [reply pkg/validation.go]: It's out of the scope of this issue. You're welcome to create ticket for this ;)

File: pkg/file.go (hunk 2 above)
Comment (dbadura) [reply pkg/file.go]: Uff I managed :D

**Comment by dbadura** [comment]
Here is test PR on test infra:
https://github.com/kyma-project/test-infra/pull/3371

All jobs are green:
https://status.build.kyma-project.io/?job=dbadura*

File: pkg/validation.go (hunk 5 above)
Comment (tgorgol) [reply pkg/validation.go]: Fair enough :)

File: pkg/validation.go (hunk 7 above)
Comment (tgorgol) [reply pkg/validation.go]: Well, I wouldn't say it's out of the scope of the issue but if it's too much work and we aren't sure how much value it will bring I agree that it's not necessary to implement it now.

**Review by tgorgol — APPROVED** [review]
LGTM
